<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="hand_O_model">

 	<xacro:macro name="base_color">
 		<color rgba="0.1 0.1 0.1 1"/>
 	</xacro:macro>

 	<xacro:macro name="swivel_color">
 		<color rgba="0.1 0.1 0.1 1"/>
 	</xacro:macro>

 	<xacro:macro name="proximal_color">
 		<color rgba="0.16 0.16 0.16 1"/>
 	</xacro:macro>

 	 <xacro:macro name="distal_color">
 		<color rgba="0.16 0.16 0.16 1"/>
 	</xacro:macro>

 	<xacro:macro name="flex_color">
 		<color rgba="0.2 0.2 .2 1"/>
 	</xacro:macro>

  <xacro:macro name="pad_color">
    <color rgba="0.106 0.227 0.631 1"/>
  </xacro:macro>  

  <xacro:property name="rubber_friction_coefficient" value="1.16" />
  <xacro:property name="rubber_contact_stiffness_coefficient" value="1000000.0" />
  <xacro:property name="rubber_contact_damping_coefficient" value="1.0" />

  <link name="world"/>

  <link name="rail">
      <visual>
          <origin rpy="0 0 0" xyz="0 0 0"/>
          <geometry>
            <cylinder length="0.8" radius="0.02"/>
          </geometry>
          <material
            name="">
            <xacro:base_color/>
          </material>
      </visual>
      <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
    </link>
    <gazebo reference="rail">
      <selfCollide>false</selfCollide>
      <material>Gazebo/White</material>
    </gazebo>

  <link name="base_link">
      <visual>
          <origin rpy="1.57 0 -1.57" xyz="0.06 0.017 0"/>
          <geometry>
            <mesh filename="package://hand_simulator/meshes/base_link.STL"/>
          </geometry>
          <material
            name="">
            <xacro:base_color/>
          </material>
      </visual>
      <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
      <!-- <collision>
        <origin rpy="1.57 0 -1.57" xyz="0.06 0.017 0"/>
        <geometry>
          <mesh filename="package://hand_simulator/meshes/base_link.STL"/>
        </geometry>
      </collision> -->
    </link>
    <gazebo reference="base_link">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Black</material>
    </gazebo>

  <link name="finger_1_1">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="1.57 3.14159 -1.57" xyz="-0.00 0.0 0.00" />
      <geometry>
        <!-- <cylinder length="0.002" radius="0.018" /> -->
        <mesh filename="package://hand_simulator/meshes/swivel_1.STL"/>
      </geometry>
    </visual>
    <!-- <collision>
      <origin rpy="0 1.57 0 " xyz="0.00 0.0 0.00" />
      <geometry>
        <cylinder length="0.002" radius="0.018" />
      </geometry>
    </collision> -->
  </link>
  <gazebo reference="finger_1_1">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_1_2">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="finger_1_2">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_1_3">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
    </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="finger_1_3">
      <selfCollide>true</selfCollide>
      <mu1>${rubber_friction_coefficient}</mu1>
      <mu2>${rubber_friction_coefficient}</mu2>
      <kp>${rubber_contact_stiffness_coefficient}</kp>
      <kd>${rubber_contact_damping_coefficient}</kd>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_2_1">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="1.57 3.14159 -1.57" xyz="-0.00 0.0 0.00" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/swivel_1.STL"/>
      </geometry>
    </visual>
    <!-- <collision>
      <origin rpy="0 1.57 0 " xyz="0.00 0.0 0.00" />
      <geometry>
        <cylinder length="0.002" radius="0.018" />
      </geometry>
    </collision> -->
  </link>
  <gazebo reference="finger_2_1">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_2_2">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
      </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="finger_2_2">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_2_3">
    <inertial>
      <origin xyz="-0.015875 -0.009000 -0.071550" />
      <mass value="1.0" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="finger_2_3">
      <selfCollide>true</selfCollide>
      <mu1>${rubber_friction_coefficient}</mu1>
      <mu2>${rubber_friction_coefficient}</mu2>
      <kp>${rubber_contact_stiffness_coefficient}</kp>
      <kd>${rubber_contact_damping_coefficient}</kd>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_3_2">
    <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.1"/>
          <inertia
            ixx="0.0" ixy="0.0" ixz="0.0"
            iyy="0.0" iyz="0.0"
            izz="0.0"/>
    </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 -3.141592 2.4" xyz="-0.00 0.0753 0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/proximal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="finger_3_2">
      <selfCollide>true</selfCollide>
      <material>Gazebo/Red</material>
    </gazebo>

  <link name="finger_3_3">
    <inertial>
      <origin xyz="-0.015875 -0.009000 -0.071550" />
      <mass value="1.0" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0" />
      </material>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -2.4962" xyz="-0.0179 0.0459 -0.008" />
      <geometry>
        <mesh filename="package://hand_simulator/meshes/distal_O.STL" scale="0.00100 0.00100 0.00100" />
      </geometry>
    </collision>
  </link> 
  <gazebo reference="finger_3_3">
      <selfCollide>true</selfCollide>
      <mu1>${rubber_friction_coefficient}</mu1>
      <mu2>${rubber_friction_coefficient}</mu2>
      <kp>${rubber_contact_stiffness_coefficient}</kp>
      <kd>${rubber_contact_damping_coefficient}</kd>
      <material>Gazebo/Red</material>
    </gazebo>-->

  <!-- Joints -->

  <joint name="world_to_rail" type="fixed">
      <parent link="world"/>
      <child link="rail"/>
      <origin
        xyz="0.37 0 0.35"
        rpy="0 0 0" />
  </joint>

  <joint name="rail_to_base" type="prismatic">
    <parent link="rail"/>
    <child link="base_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="-100." upper="100." velocity="4"/>
    <origin
      xyz="-0.06 0 0.42"
      rpy="-1.57 0 0" />
    <dynamics damping="10"/>
  </joint>

 <joint name="base_to_finger_1_1" type="revolute">
   <origin rpy="3.14159 0 0" xyz="-0.003 -0.034 -0.027" />
   <parent link="base_link" />
   <child link="finger_1_1" />
   <limit effort="1000" lower="0" upper="1.57" velocity="200" />
   <axis xyz="1 0 0" />
   <dynamics damping="5" friction="0.0" />
 </joint>

 <joint name="finger_1_1_to_finger_1_2" type="revolute">
   <origin rpy="0 0 0" xyz="-0.02 0 0" />
   <parent link="finger_1_1" />
   <child link="finger_1_2" />
   <limit effort="1000.0" lower="-0.1" upper="3.0" velocity="10"/>
   <axis xyz="0 0 1" />
   <dynamics damping="5" friction="0.0" />
 </joint>

 <joint name="finger_1_2_to_finger_1_3" type="revolute">
   <origin rpy="0 0 0" xyz="0.000800 0.062 0" />
   <parent link="finger_1_2" />
   <child link="finger_1_3" />
   <limit effort="1000.0" lower="0.0" upper="1.27" velocity="10"/>
   <axis xyz="0 0 1" />
   <dynamics damping="5" friction="0.0" />
  </joint>

  <joint name="base_to_finger_2_1" type="revolute">
    <origin rpy="3.14159 0 0" xyz="-0.003 -0.034 0.027" />
    <parent link="base_link" />
    <child link="finger_2_1" />
    <limit effort="1000" lower="-1.57" upper="0" velocity="200" />
    <axis xyz="1 0 0" />
    <dynamics damping="5" friction="0.0" />
  </joint>

  <joint name="finger_2_1_to_finger_2_2" type="revolute">
    <origin rpy="0 0 0" xyz="-0.02 0 0" />
    <parent link="finger_2_1" />
    <child link="finger_2_2" />
    <limit effort="1000" lower="-0.1" upper="3.0" velocity="10" />
    <axis xyz="0 0 1" />
    <dynamics damping="5" friction="0.0" />
  </joint>

  <joint name="finger_2_2_to_finger_2_3" type="revolute">
    <origin rpy="0 0 0" xyz="0.000800 0.062 0" />
    <parent link="finger_2_2" />
    <child link="finger_2_3" />
    <limit effort="1000" lower="0.0" upper="1.27" velocity="10" />
    <axis xyz="0 0 1" />
    <dynamics damping="5" friction="0.0" />
  </joint>

  <joint name="base_to_finger_3_2" type="revolute">
    <origin rpy="0 0 0" xyz="-0.023 0.034 0.000000" />
    <parent link="base_link" />
    <child link="finger_3_2" />
    <limit effort="1000" lower="-0.1" upper="3.0" velocity="10" />
    <axis xyz="0 0 1" />
    <dynamics damping="5" friction="0.0" />
  </joint>

  <joint name="finger_3_2_to_finger_3_3" type="revolute">
    <origin rpy="0 0 0" xyz="0.000800 0.062 0" />
    <parent link="finger_3_2" />
    <child link="finger_3_3" />
    <limit effort="1000" lower="0.0" upper="1.27" velocity="10" />
    <axis xyz="0 0 1" />
    <dynamics damping="5" friction="0.0" />
  </joint>


 <!-- Transmissions -->

  <transmission name="lift" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$lift_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="rail_to_base">
          <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger1_1_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger11_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="base_to_finger_1_1">
          <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger1_2_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger12_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="finger_1_1_to_finger_1_2">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger1_3_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger13_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="finger_1_2_to_finger_1_3">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger2_1_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger21_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="base_to_finger_2_1">
          <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger2_2_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger22_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="finger_2_1_to_finger_2_2">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger2_3_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger23_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="finger_2_2_to_finger_2_3">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger2_3_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger32_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="base_to_finger_3_2">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission>

  <transmission name="finger3_3_trans" type="SimpleTransmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$finger33_motor">
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="finger_3_2_to_finger_3_3">
          <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
  </transmission> 
  
    <gazebo>
      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
          <robotNamespace>/hand</robotNamespace>
      </plugin>
    </gazebo>

</robot>
